src/Elm/Syntax/DestructurPattern.elm

```diff
@@ -0,0 +1,210 @@
+module Elm.Syntax.DestructurPattern exposing
+    ( QualifiedNameRef
+    , moduleNames
+    , encode, decoder
+    , DestructurPattern(..)
+    )
+
+{-|
+
+
+# Pattern Syntax
+
+This syntax represents the patterns.
+For example:
+
+    Just x as someMaybe
+    {name, age}
+
+
+# Types
+
+@docs Pattern, QualifiedNameRef
+
+
+## Functions
+
+@docs moduleNames
+
+
+## Serialization
+
+@docs encode, decoder
+
+-}
+
+import Elm.Json.Util exposing (decodeTyped, encodeTyped)
+import Elm.Syntax.ModuleName as ModuleName exposing (ModuleName)
+import Elm.Syntax.Node as Node exposing (Node(..))
+import Json.Decode as JD exposing (Decoder)
+import Json.Encode as JE exposing (Value)
+
+
+{-| Custom type for all patterns such as:
+
+  - `AllPattern`: `_`
+  - `UnitPattern`: `()`
+  - `CharPattern`: `'c'`
+  - `StringPattern`: `"hello"`
+  - `IntPattern`: `42`
+  - `HexPattern`: `0x11`
+  - `FloatPattern`: `42.0`
+  - `TuplePattern`: `(a, b)`
+  - `RecordPattern`: `{name, age}`
+  - `UnConsPattern`: `x :: xs`
+  - `ListPattern`: `[ x, y ]`
+  - `VarPattern`: `x`
+  - `NamedPattern`: `Just _`
+  - `AsPattern`: `_ as x`
+  - `ParenthesizedPattern`: `( _ )`
+
+-}
+type DestructurPattern
+    = AllPattern_
```

Comment:
I chose to add underscores to all the variants so there isn't any name collisions with `Pattern`.

Reply:
I understand. For now I think it is fine. Personally I'm not really fond of `_` I think i I'm planning to renaming all the pattern stuff. Eliminate the `Pattern` in the value constructors and 'enforce' reference by module name:
`Pattern.All` and `DestructurAll`.